### DISCUSSION TITLE
Set guess control pulses

### DISCUSSION THREAD

**Discussion by Leeseok-628** [qucontrol/krotov#101] [Q&A] [answered]
Hi,

I want to start a new optimization with a guess control field as the previously optimized control field. Is it possible? 

Equivalently, is there any way to start my guess control with specific values?

**Comment by goerz** [comment, marked as answer]
Yes, both of these things are discussed in the documentation!

* [How to continue from a previous optimization](https://qucontrol.github.io/krotov/v1.2.1/10_howto.html#how-to-continue-from-a-previous-optimization)
* [Controls can be arrays instead of functions](https://qucontrol.github.io/krotov/v1.2.1/08_qutip_usage.html?highlight=array%20control%20values)

**Reply by Leeseok-628** [reply]
Thank you for the response!

I actually already tried the first one. After my first optimization of J = <ϕ(T)|O|ϕ(T)> is done (stored in opt_result), I only changed the operator O and fixed other parameters the same. Then I added `continue_from=opt_result` to `krotov.optimize_pulses`, but it doens't iterate at all (even though I set iter_stop=10) and just terminates without being optimized.

**Reply by goerz** [reply]
The iterations keep counting, so if your first optimization did 10 iterations, you'd have to set `iter_stop` to something higher

**Reply by Leeseok-628** [reply]
I've been using different shape functions S(t) to optimize the same problem, and some particular sets of shape functions give much faster convergence rate! (and of course, the optimized pulse amplitudes look very different as well, smoother and more physical) Is there any literature on this aspect so that I can get some intuitions/sense on how to cleverly choose the shape functions for different problems (i.e. different J and/or J_T)? (Perhaps, setting the shape functions similar to the guess controls would help..)

**Reply by goerz** [reply]
This is expected. It very much depends on the exact system. There is no theory behind it. It pays off to play around a little, both for the update shape and even more for the guess pulse directly. This is why optimal control can be a bit of an art.

**Reply by Leeseok-628** [reply]
Got it, thanks. Also, how do we deal with the Hamiltonian with nonlinear control fields (e.g. H = cos(e_1(t))*X + cos(e_2(t))*Z)? It seems like we have to calculate proper σ(t) for the second-order update, and I have to follow the calculation in equation 13, 14 in https://arxiv.org/pdf/1008.5126.pdf, which seems little unclear to me on how to choose appropriate parameters A, B, and C (or ε_i).

**Reply by christianekoch** [reply]
You don't need the second order for a non-linear dependence on the control, so A, B and C can still be zero if your target functional and equations of motion allow for that. The relevant equation to ensure monotonic convergence for non-linear dependencies on the control are (32) and (30) in the preprint you linked.

**Reply by goerz** [reply]
For nonlinear controls you may have to increase λₐ a bit. This is because technically, the value of the *updated* field will appear in the right-hand-side of the update equation. We deal that with by plugging in the value of the *guess* field instead, which implies small updates, or large λₐ.

**Reply by Leeseok-628** [reply]
I read through the paper step by step. Thanks for the comments :)

Suppose we have H=H0+e(t)Hx. When e(t) is non-linear such that e(t) = cos(t), I can simply set Hamiltonian as `H = [H0, [Hx, (lambda t : np.cos(t)]]`. This will optimize e(t) for each time grid. When e(t) is linear as e(t) = a*t, then `H = [H0, [Hx, (lambda t: a*t)]]` for given a. For any case, the initial pulse is equal to e(t) (i.e. cos(t) and a*t respectively) for t in the given time grid. After the optimization is done, the optimized pulses might look pretty different from the initial pulse, which sometimes look non-physical. 

So I'm wondering if there's any way to optimize the pulse while perfectly maintaining its function form so that the optimized pulse look more physical. For instance, is it possible to optimize for A, w, and Φ in e(t) = A*cos(wt + Φ), so that the optimized pulse in every iterations maintain as cosine waves (but with different frequency, amplitudes, and phase)?

**Comment by goerz** [comment]
> Suppose we have H=H0+e(t)Hx. When e(t) is non-linear such that e(t) = cos(t), I can simply set Hamiltonian as `H = [H0, [Hx, (lambda t : np.cos(t)]]`. This will optimize e(t) for each time grid. When e(t) is linear as e(t) = a_t, then `H = [H0, [Hx, (lambda t: a*t)]]` 
for given a.

When we talk about linear, we mean linear in `e`, not linear in `t`. `H=H0+e(t)Hx` is linear for any `e(t)`. A non-linear (quadratic) control would be `H=H0+e²(t)Hx`.

> After the optimization is done, the optimized pulses might look pretty different from the initial pulse, which sometimes look non-physical.

What do you mean by non-physical? If it's too "spiky", increase λₐ.

> So I'm wondering if there's any way to optimize the pulse while perfectly maintaining its function form so that the optimized pulse look more physical. For instance, is it possible to optimize for A, w, and Φ in e(t) = A*cos(wt + Φ), so that the optimized pulse in every iterations maintain as cosine waves (but with different frequency, amplitudes, and phase)?

Not with Krotov's method, no. If you have just a handful of free parameters, you can use [gradient-free optimization](https://qucontrol.github.io/krotov/v1.2.1/11_other_methods.html#gradient-free-optimization). You might also be interested in [GROUP/GOAT](https://qucontrol.github.io/krotov/v1.2.1/11_other_methods.html?highlight=group%20goat). Krotov's method targets mainly unconstrained controls `e(t)` (up to the restrictions implied by the `update_shape`). There are a few tricks to add some amplitude and spectral constraints, but these are already pushing the method. If you have more fundamental constraints, e.g., only allowing specific control functions, you have to use a different method.

**Reply by Leeseok-628** [reply]
I see, great, thank you. Can [GRAPE](https://qutip.org/docs/4.0.2/guide/guide-control.html#the-grape-algorithm) also do such task as well?

**Reply by goerz** [reply]
No, GRAPE is pretty similar to Krotov's method in this respect. GROUP/GOAT are in some sense variations of GRAPE that deal with the kind of analytical pulses you have in mind. Did you see our discussion on [how to choose optimization methods](https://qucontrol.github.io/krotov/v1.2.1/11_other_methods.html#choosing-an-optimization-method)?
